tensorflow/tools/pip_package/setup.py

```diff
@@ -77,8 +77,9 @@
 REQUIRED_PACKAGES = [
     # NOTE: As numpy has releases that break semver guarantees and several other
```

Comment:
For maintainers: this comment lacks context. Nothing in Python uses semver, including Python itself and TensorFlow. The guidance for NumPy's versioning scheme is at https://numpy.org/devdocs/user/depending_on_numpy.html#understanding-numpy-s-versioning-and-api-abi-stability. The correct upper bounds where you do get backwards compat guarantees is "two minor versions into the future", so right now that is `<1.24.0`.

Reply:
You are right, the versioning scheme is similar to semver (compared to PVP for example) but the backwards compat guarantees are less severe.

We have a roadmap at the end of which we would be able to use a wide range of dependencies in `setup.py` after testing that TF works with any dependency version in the range. But this would take a significant amount of work that has just recently been prioritized